### PR TITLE
fix: support update of description for project api key resource

### PR DESCRIPTION
**Pull request by AgustinBettati**
## Description

Ticket: [INTMDB-917](https://jira.mongodb.org/browse/INTMDB-917)

Updating description of an api key in the mongodbatlas_project_api_key resource was not contemplated in update. This includes a new request for handling this scenario.

## Type of change:

- [x] Bug fix (non-breaking change which fixes an issue). Please, add the "bug" label to the PR.
- [ ] New feature (non-breaking change which adds functionality). Please, add the "enhancement" label to the PR.
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected). Please, add the "breaking change" label to the PR.
- [ ] This change requires a documentation update
- [x] Documentation fix/enhancement

## Required Checklist:

- [x] I have signed the [MongoDB CLA](https://www.mongodb.com/legal/contributor-agreement)
- [x] I have read the [contribution guidelines](https://github.com/mongodb/terraform-provider-mongodbatlas/blob/master/CONTRIBUTING.md)
- [x] I have added tests that prove my fix is effective or that my feature works per HashiCorp requirements
- [x] I have added any necessary documentation (if appropriate)
- [x] I have run make fmt and formatted my code

## Further comments
- Found that api_project_api_key documentation had some inconsistencies, took the opportunity to adjust them here.

### CODE REVIEW

File: website/docs/r/project_api_key.html.markdown [hunk 1]
```diff
@@ -57,24 +57,19 @@ resource "mongodbatlas_project_api_key" "test" {
...
+* `role_names` - (Required) List of Project roles that the Programmatic API key needs to have. Ensure you provide: at least one role and ensure all roles are valid for the Project.  You must specify an array even if you are only associating a single role with the Programmatic API key.
 The following are valid roles:
   * `GROUP_OWNER`
   * `GROUP_READ_ONLY`
   * `GROUP_DATA_ACCESS_ADMIN`
   * `GROUP_DATA_ACCESS_READ_WRITE`
   * `GROUP_DATA_ACCESS_READ_ONLY`
-  * `GROUP_CLUSTER_MANAGER`  
-  * 
-### Programmatic API Keys
```
Comment (AgustinBettati) [inline website/docs/r/project_api_key.html.markdown:73]: This section is unrelated to project api key resource, seems like a copy paste that came from project resource documentation

File: website/docs/r/project_api_key.html.markdown [hunk 2]
```diff
@@ -57,24 +57,19 @@ resource "mongodbatlas_project_api_key" "test" {
 
 ~> **NOTE:** Project created by API Keys must belong to an existing organization.
 
-# project_assignment
+### project_assignment
 Project Assignment attribute is optional (Use project_assignment going forward as role_names parameter above is deprecated)
 
 * `project_id` - (Required) Project ID to assign to Access Key
-* `role_names` - Name of the role. This resource returns all the roles the user has in Atlas.
+* `role_names` - (Required) List of Project roles that the Programmatic API key needs to have. Ensure you provide: at least one role and ensure all roles are valid for the Project.  You must specify an array even if you are only associating a single role with the Programmatic API key.
```
Comment (AgustinBettati) [inline website/docs/r/project_api_key.html.markdown]: role_names is marked as required in our schema, and also verified that an empty array causes project assignment to not be created.

File: mongodbatlas/resource_mongodbatlas_project_api_key.go [hunk 3]
```diff
@@ -225,12 +226,23 @@ func resourceMongoDBAtlasProjectAPIKeyRead(ctx context.Context, d *schema.Resour
...
 func resourceMongoDBAtlasProjectAPIKeyUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
 	conn := meta.(*MongoDBClient).Atlas
+	connV2 := meta.(*MongoDBClient).AtlasV2
+
 	ids := decodeStateID(d.Id())
 	projectID := ids["project_id"]
 	apiKeyID := ids["api_key_id"]
 
-	updateRequest := new(matlas.AssignAPIKey)
+	if d.HasChange("description") {
```
Comment (wtrocki) [inline mongodbatlas/resource_mongodbatlas_project_api_key.go]: nit: Failure in update of description block assign. Maybe description update can be after assignment?

**Review by wtrocki — APPROVED** [review]

File: website/docs/r/project_api_key.html.markdown [hunk 4]
```diff
@@ -57,24 +57,19 @@ resource "mongodbatlas_project_api_key" "test" {
...
   * `GROUP_OWNER`
   * `GROUP_READ_ONLY`
   * `GROUP_DATA_ACCESS_ADMIN`
   * `GROUP_DATA_ACCESS_READ_WRITE`
   * `GROUP_DATA_ACCESS_READ_ONLY`
-  * `GROUP_CLUSTER_MANAGER`  
-  * 
-### Programmatic API Keys
-api_keys allows one to assign an existing organization programmatic API key to a Project. The api_keys attribute is optional.
+  * `GROUP_CLUSTER_MANAGER`
```
Comment (andreaangiolillo) [inline website/docs/r/project_api_key.html.markdown]: can we remove this list of role and add a link to the documentation? https://www.mongodb.com/docs/atlas/reference/api-resources-spec/v2/#tag/Programmatic-API-Keys/operation/addProjectApiKey

Comment (andreaangiolillo) [reply website/docs/r/project_api_key.html.markdown]: same in other part of the file and in the data source documentation

File: mongodbatlas/resource_mongodbatlas_project_api_key.go [hunk 5]
```diff
@@ -225,12 +226,23 @@ func resourceMongoDBAtlasProjectAPIKeyRead(ctx context.Context, d *schema.Resour
...
+		newDescription := d.Get("description").(string)
+		_, _, err := connV2.ProgrammaticAPIKeysApi.UpdateApiKeyRoles(ctx, projectID, apiKeyID, &admin.CreateAtlasProjectApiKey{
+			Desc: &newDescription,
+		}).Execute()
+		if err != nil {
+			return diag.Errorf("error updating description in api key(%s): %s", apiKeyID, err)
+		}
```
Comment (andreaangiolillo) [inline mongodbatlas/resource_mongodbatlas_project_api_key.go]:
```suggestion
		if _, _, err := connV2.ProgrammaticAPIKeysApi.UpdateApiKeyRoles(ctx, projectID, apiKeyID, &admin.CreateAtlasProjectApiKey{
			Desc: &newDescription,
		}).Execute(); err != nill {
			return diag.Errorf("error updating description in api key(%s): %s", apiKeyID, err)
		}
```

**Review by andreaangiolillo — APPROVED** [review]
LGTM. My comments are not blocking the PR. Great work!

File: mongodbatlas/resource_mongodbatlas_project_api_key.go (hunk 3 above)
Comment (AgustinBettati) [reply mongodbatlas/resource_mongodbatlas_project_api_key.go]: agree, will adjust

File: website/docs/r/project_api_key.html.markdown (hunk 4 above)
Comment (AgustinBettati) [reply website/docs/r/project_api_key.html.markdown]: good point, adjusted. Used the following link https://www.mongodb.com/docs/atlas/reference/user-roles/#project-roles for referencing

**Comment by AgustinBettati** [comment]
Hello @pierwill @zach-carr @jwilliams-mongo! I will be merging this PR, but feel free to review and if you have any comments I can address them in a separate PR.
